### PR TITLE
let bootstrap takes in optional refs (for gerrit)

### PR DESCRIPTION
**Pull request by krzyzacy**
/hold
for discussion

So basically I'm trying to resolve https://github.com/kubernetes/test-infra/issues/7426, this works but not super great - because currently bootstrap assumes github pull refs and generates that from the PR number.

How we should handle this better, in podutils? I'm also thinking of add a `Ref` field in `kube.Refs`, and we can construct `$PULL_REFS` accordingly.

/assign @BenTheElder @cjwagner @stevekuznetsov

### CODE REVIEW

**Comment by stevekuznetsov** [comment]
Hmm, we just added the path alias field to the refs, I imagine we might add pull method (https/http/git+ssh), might as well add refspec as well, although it's not used in the current clone logic, here is an example run:

```
$ git init
$ git config user.name ci-robot
$ git config user.email ci-robot@k8s.io
$ git fetch https://github.com/openshift/origin.git --tags --prune
$ git fetch https://github.com/openshift/origin.git master
$ git checkout 10927c956187c547bc64700198842f1daa1fdf28
$ git branch --force master 10927c956187c547bc64700198842f1daa1fdf28
$ git checkout master
$ git fetch https://github.com/openshift/origin.git pull/19248/head
$ git merge d6ea33b1c388c7dfc76f3f593a667d2acd013cd1
```

**Comment by stevekuznetsov** [comment]
So you can probably make the change to the config and utility in this pull too if you'd like, does `bootstrap.py` read `$JOB_SPEC`?

**Comment by krzyzacy** [comment]
not really - we pass `--repo=$(REPO_NAME)=$(PULL_REFS)` in the config

**Comment by krzyzacy** [comment]
the issue is, how do you get `pull/19248/head`? That's constructed from PULL_NUMBER?

**Comment by stevekuznetsov** [comment]
Nope, I introduced `$JOB_SPEC` so that we can deserialize JSON and just use data instead of trying to mess with shell environment variables. We clone using `kube.Refs` [here](https://github.com/kubernetes/test-infra/blob/master/prow/pod-utils/clone/clone.go#L32) and deserialize it [here](https://github.com/kubernetes/test-infra/blob/master/prow/clonerefs/options.go#L48) or optionally from user flags like `--repo` [here](https://github.com/kubernetes/test-infra/blob/master/prow/clonerefs/parse.go#L40).

**Comment by krzyzacy** [comment]
@stevekuznetsov looks like https://github.com/kubernetes/test-infra/blob/master/prow/pod-utils/clone/clone.go#L71 won't work with gerrit refs :-( we probably need to make that configurable?

**Comment by krzyzacy** [comment]
added a `Ref` field to `kube.Pull` and this might be a bit more clearer.

File: prow/kube/prowjob.go [hunk 1]
```diff
@@ -185,7 +186,13 @@ type Refs struct {
 func (r Refs) String() string {
 	rs := []string{fmt.Sprintf("%s:%s", r.BaseRef, r.BaseSHA)}
 	for _, pull := range r.Pulls {
-		rs = append(rs, fmt.Sprintf("%d:%s", pull.Number, pull.SHA))
+		ref := fmt.Sprintf("%d:%s", pull.Number, pull.SHA)
+
+		if pull.Ref != "" {
+			ref = fmt.Sprintf("%s:%s", ref, pull.Ref)
```
[the remark this answers is not in the document]
Comment (krzyzacy) [reply prow/kube/prowjob.go]: hummm, this allows `split(':')`, and PR_NUMBER, SHA, Ref should be parallel

File: prow/kube/prowjob.go [hunk 2]
```diff
@@ -164,6 +164,7 @@ type Pull struct {
 	Number int    `json:"number,omitempty"`
 	Author string `json:"author,omitempty"`
 	SHA    string `json:"sha,omitempty"`
+	Ref    string `json:"ref,omitempty"`
```
[the remark this answers is not in the document]
Comment (krzyzacy) [reply prow/kube/prowjob.go]: done

**Comment by krzyzacy** [comment]
patched up cloneref and podutils/clone

**Comment by stevekuznetsov** [comment]
/lgtm
/hold
@cjwagner @BenTheElder PTAL

**Comment by krzyzacy** [comment]
I'll deploy this to gerrit, will not affect k8s prow

(also need to :eyes: bootstrap)
/hold cancel

**Pull request merged**